src/blocks/battery/upower.rs

```diff
@@ -57,86 +57,138 @@ impl Device {
-            .error("Failed to get status")?
-        {
-            1 => BatteryStatus::Charging,
-            2 | 6 => BatteryStatus::Discharging,
-            3 => BatteryStatus::Empty,
-            4 => BatteryStatus::Full,
-            5 => BatteryStatus::NotCharging,
-            _ => BatteryStatus::Unknown,
-        };
-
-        let time_remaining = match status {
-            BatteryStatus::Charging => Some(
-                self.device_proxy
-                    .time_to_full()
-                    .await
-                    .error("Failed to get time to full")? as f64,
-            ),
-            BatteryStatus::Discharging => Some(
-                self.device_proxy
-                    .time_to_empty()
-                    .await
-                    .error("Failed to get time to empty")? as f64,
-            ),
-            _ => None,
-        };
-
-        Ok(Some(BatteryInfo {
-            status,
-            capacity,
-            power: Some(power),
-            time_remaining,
-        }))
+        match &self.device_conn {
+            None => Ok(None),
+            Some(device_conn) => {
+                match try_join! {
+                    device_conn.device_proxy.percentage(),
+                    device_conn.device_proxy.energy_rate(),
+                    device_conn.device_proxy.state(),
+                    device_conn.device_proxy.time_to_full(),
+                    device_conn.device_proxy.time_to_empty(),
+                } {
+                    Err(_) => Ok(None),
+                    Ok((capacity, power, state, time_to_full, time_to_empty)) => {
+                        let status = match state {
+                            1 => BatteryStatus::Charging,
+                            2 | 6 => BatteryStatus::Discharging,
+                            3 => BatteryStatus::Empty,
+                            4 => BatteryStatus::Full,
+                            5 => BatteryStatus::NotCharging,
+                            _ => BatteryStatus::Unknown,
+                        };
+
+                        let time_remaining = match status {
+                            BatteryStatus::Charging => Some(time_to_full as f64),
+                            BatteryStatus::Discharging => Some(time_to_empty as f64),
+                            _ => None,
+                        };
+
+                        Ok(Some(BatteryInfo {
+                            status,
+                            capacity,
+                            power: Some(power),
+                            time_remaining,
+                        }))
+                    }
+                }
+            }
+        }
     }
 
     async fn wait_for_change(&mut self) -> Result<()> {
-        self.changes.next().await;
+        match &mut self.device_conn {
+            Some(device_conn) => {
+                select! {
+                    _ = device_conn.changes.next() => {},
+                    _ = self.device_added_stream.next() => {},
```

Comment:
As is, this will trigger block updates on "device_added" event even though it is irrelevant. So the entire `select!` must be wrapped in a `loop` which will break only when a relevant event is received.